lib/graphql/client/http.rb

```diff
@@ -87,6 +97,48 @@ def connection
           client.use_ssl = uri.scheme == "https"
         end
       end
+
+    private
+
+      # generate the form data for a multipart request according to the GraphQL multipart request
+      # specification (https://github.com/jaydenseric/graphql-multipart-request-spec/)
+      #
+      # Example request form data:
+      #   operations: {"query": "…", "operationName": "addToGallery", "variables": {"galleryId": "…", images: [null, null, null]}}
+      #   map: {"1": ["variables.images.0"], "2": ["variables.images.1"], "3": ["variables.images.2"]}
+      #   1: File
+      #   2: File
+      #   3: File
+      #
+      # note: modifies `variables`, returns form data (except `operations`) or `nil`
+      def form_data!(variables)
+        form_data = []
+        file_map = {}
+
+        # recursively walk `variables` looking for `File` values, add them to the form data,
+        # then replace with `nil`
+        stack = variables.map { |k, v|  [variables, ["variables", k], v] }
+        while (variable, path, val = stack.pop) do
+          if val.is_a?(Hash)
+            val.each { |k, v|  stack.push [val, path.dup << k, v] }
+
+          elsif val.is_a?(Array)
+            val.each.with_index { |v, i|  stack.push [val, (path.dup << i), v] }
+
+          elsif val.respond_to?(:to_path)
+            metadata = { filename: File.basename(val.to_path) }
+            metadata[:content_type] = val.content_type  if val.respond_to?(:content_type)
+
+            idx = file_map.length + 1
+            file_map[idx.to_s] = [path.map(&:to_s).join(".")]
+            form_data.append [idx.to_s, val, metadata]
+
+            variable[path.last] = nil  # replace `IO` value with `nil` in `variables`
+          end
+        end
+
+        form_data.presence && form_data.append(["map", JSON.generate(file_map)])
```

Comment:
see https://github.com/github/graphql-client/pull/236/files#r1280049102

```suggestion
        form_data.presence && form_data.prepend(["map", JSON.generate(file_map)])
```